Implement basic field assertion (types, presence)

mtojek opened the pull request:
This PR introduces a basic field type assertion during pipeline and system tests. When a pipeline or system test is executed, generated results are examined against correctness with fields schema (fields YAML files). All inconsistencies will be reported.

Issue: https://github.com/elastic/elastic-package/issues/90

Action items (TODOs):

- [x] Add AWS sample data stream with fields
- [x] Load JSON unit test (with ES document) from JSON file
- [x] Validate JSON results in pipeline tests
- [x] Validate JSON results in system tests
- [x] Validate document body
- [x] Validate field type
- [x] Fix all edge cases in pipeline tests
- [x] Fix all edge cases in system tests

ycombinator commented:
I assume you will choose a different test package than `aws` when it comes to validating JSON body in system tests?

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Felastic-package%2FPR-143/detail/PR-143/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Felastic-package%2FPR-143/detail/PR-143/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Felastic-package%2FPR-143/detail/PR-143/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Felastic-package%2FPR-143/detail/PR-143/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://elastic-package_143.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `[Pull request #143 opened]`

* Start Time: 2020-10-19T14:25:35.483+0000

* Duration: 8 min 14 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->


### Steps errors [![1](https://img.shields.io/badge/1%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Felastic-package%2FPR-143/detail/PR-143/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>
  
  
  * **Name**: `Check`
    * Description: make check
    
    * Duration: 0 min 29 sec
    
    * Start Time: 2020-10-19T14:33:18.595+0000
    
    * [log](https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/elastic-package/pipelines/PR-143/runs/1/steps/114/log)
    
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2020-10-19T14:32:57.125Z] + git rev-parse HEAD
[2020-10-19T14:32:57.500Z] + git rev-parse HEAD
[2020-10-19T14:32:57.815Z] + git rev-parse origin/pr/143
[2020-10-19T14:32:57.855Z] [INFO] githubEnv: Found Git Build Cause: pr
[2020-10-19T14:32:58.266Z] Masking supported pattern matches of $GITHUB_TOKEN
[2020-10-19T14:32:59.150Z] [WARN] githubApiCall: The REST API call https://api.github.com/repos/elastic/elastic-package/pulls/143/reviews return 0 elements
[2020-10-19T14:32:59.196Z] [INFO] githubPrCheckApproved: Title: Implement field assertion - User: mtojek - Author Association: MEMBER
[2020-10-19T14:33:00.308Z] Stashed 440 file(s)
[2020-10-19T14:33:00.384Z] Running in /var/lib/jenkins/workspace/Beats_elastic-package_PR-143/src/github.com/elastic/elastic-package
[2020-10-19T14:33:01.325Z] + gvm 1.14.2
[2020-10-19T14:33:07.927Z] export GOROOT="/var/lib/jenkins/workspace/Beats_elastic-package_PR-143/.gvm/versions/go1.14.2.linux.amd64"
[2020-10-19T14:33:07.927Z] export PATH="/var/lib/jenkins/workspace/Beats_elastic-package_PR-143/.gvm/versions/go1.14.2.linux.amd64/bin:$PATH"
[2020-10-19T14:33:08.294Z] + go get -u github.com/magefile/mage
[2020-10-19T14:33:10.587Z] + go get -u github.com/elastic/go-licenser
[2020-10-19T14:33:11.894Z] + go get -u golang.org/x/tools/cmd/goimports
[2020-10-19T14:33:17.579Z] + go get -u github.com/jstemmer/go-junit-report
[2020-10-19T14:33:18.577Z] Running in /var/lib/jenkins/workspace/Beats_elastic-package_PR-143/src/github.com/elastic/elastic-package
[2020-10-19T14:33:18.889Z] + make check
[2020-10-19T14:33:18.889Z] go get -ldflags "-X github.com/elastic/elastic-package/internal/version.CommitHash=`git describe --always --long --dirty` -X github.com/elastic/elastic-package/internal/version.BuildTime=`date +%s`" \
[2020-10-19T14:33:18.889Z]     github.com/elastic/elastic-package
[2020-10-19T14:33:19.836Z] go: downloading github.com/pkg/errors v0.9.1
[2020-10-19T14:33:19.836Z] go: downloading github.com/go-git/go-git/v5 v5.1.0
[2020-10-19T14:33:19.836Z] go: downloading github.com/elastic/package-spec/code/go v0.0.0-20201014114156-e7102f8cbd39
[2020-10-19T14:33:19.836Z] go: downloading github.com/AlecAivazis/survey/v2 v2.1.1
[2020-10-19T14:33:19.836Z] go: downloading gopkg.in/yaml.v3 v3.0.0-20200615113413-eeeca48fe776
[2020-10-19T14:33:19.836Z] go: downloading github.com/google/go-github/v32 v32.1.0
[2020-10-19T14:33:19.836Z] go: downloading github.com/Masterminds/semver v1.5.0
[2020-10-19T14:33:19.836Z] go: downloading github.com/go-git/go-billy/v5 v5.0.0
[2020-10-19T14:33:19.837Z] go: downloading github.com/magefile/mage v1.10.0
[2020-10-19T14:33:19.837Z] go: downloading github.com/elastic/go-elasticsearch/v7 v7.9.0
[2020-10-19T14:33:19.837Z] go: downloading github.com/elastic/go-ucfg v0.8.3
[2020-10-19T14:33:19.837Z] go: downloading golang.org/x/crypto v0.0.0-20200820211705-5c72a883971a
[2020-10-19T14:33:19.837Z] go: downloading github.com/spf13/cobra v1.0.0
[2020-10-19T14:33:19.837Z] go: downloading github.com/creasty/defaults v1.5.1
[2020-10-19T14:33:19.837Z] go: downloading github.com/kballard/go-shellquote v0.0.0-20180428030007-95032a82bc51
[2020-10-19T14:33:20.099Z] go: downloading github.com/aymerick/raymond v2.0.2+incompatible
[2020-10-19T14:33:20.099Z] go: downloading github.com/kylelemons/godebug v1.1.0
[2020-10-19T14:33:20.099Z] go: downloading github.com/coreos/pkg v0.0.0-20180928190104-399ea9e2e55f
[2020-10-19T14:33:20.099Z] go: downloading github.com/xeipuuv/gojsonschema v1.2.0
[2020-10-19T14:33:20.099Z] go: downloading github.com/cespare/xxhash v1.1.0
[2020-10-19T14:33:20.099Z] go: downloading github.com/mgutz/ansi v0.0.0-20200706080929-d51e80ef957d
[2020-10-19T14:33:20.099Z] go: downloading github.com/rakyll/statik v0.1.7
[2020-10-19T14:33:20.099Z] go: downloading github.com/cespare/xxhash/v2 v2.1.1
[2020-10-19T14:33:20.099Z] go: downloading github.com/go-git/gcfg v1.5.0
[2020-10-19T14:33:20.099Z] go: downloading github.com/mattn/go-isatty v0.0.12
[2020-10-19T14:33:20.099Z] go: downloading golang.org/x/text v0.3.3
[2020-10-19T14:33:20.099Z] go: downloading github.com/sergi/go-diff v1.1.0
[2020-10-19T14:33:20.099Z] go: downloading gopkg.in/yaml.v2 v2.3.0
[2020-10-19T14:33:20.099Z] go: downloading github.com/xeipuuv/gojsonreference v0.0.0-20180127040603-bd5ef7bd5415
[2020-10-19T14:33:20.099Z] go: downloading golang.org/x/sys v0.0.0-20200915084602-288bc346aa39
[2020-10-19T14:33:20.099Z] go: downloading github.com/jbenet/go-context v0.0.0-20150711004518-d14ea06fba99
[2020-10-19T14:33:20.099Z] go: downloading github.com/emirpasic/gods v1.12.0
[2020-10-19T14:33:20.099Z] go: downloading github.com/Masterminds/semver/v3 v3.1.0
[2020-10-19T14:33:20.099Z] go: downloading github.com/google/go-querystring v1.0.0
[2020-10-19T14:33:20.099Z] go: downloading github.com/mitchellh/go-homedir v1.1.0
[2020-10-19T14:33:20.099Z] go: downloading github.com/inconshreveable/mousetrap v1.0.0
[2020-10-19T14:33:20.099Z] go: downloading github.com/mattn/go-colorable v0.1.7
[2020-10-19T14:33:20.361Z] go: downloading gopkg.in/warnings.v0 v0.1.2
[2020-10-19T14:33:20.361Z] go: downloading github.com/xeipuuv/gojsonpointer v0.0.0-20190905194746-02993c407bfb
[2020-10-19T14:33:20.361Z] go: downloading golang.org/x/oauth2 v0.0.0-20200902213428-5d25da1a8d43
[2020-10-19T14:33:20.361Z] go: downloading github.com/spf13/pflag v1.0.5
[2020-10-19T14:33:20.361Z] go: downloading github.com/imdario/mergo v0.3.11
[2020-10-19T14:33:20.361Z] go: downloading golang.org/x/net v0.0.0-20200904194848-62affa334b73
[2020-10-19T14:33:20.361Z] go: downloading github.com/kevinburke/ssh_config v0.0.0-20190725054713-01f96b0aa0cd
[2020-10-19T14:33:20.361Z] go: downloading github.com/xanzy/ssh-agent v0.2.1
[2020-10-19T14:33:20.361Z] go: downloading google.golang.org/appengine v1.6.6
[2020-10-19T14:33:20.624Z] go: downloading github.com/golang/protobuf v1.4.2
[2020-10-19T14:33:20.624Z] go: downloading google.golang.org/protobuf v1.25.0
[2020-10-19T14:33:42.636Z] go get -u golang.org/x/tools/cmd/goimports
[2020-10-19T14:33:42.636Z] go: downloading golang.org/x/tools v0.0.0-20201017001424-6003fad69a88
[2020-10-19T14:33:42.637Z] go: found golang.org/x/tools/cmd/goimports in golang.org/x/tools v0.0.0-20201017001424-6003fad69a88
[2020-10-19T14:33:42.637Z] go: downloading golang.org/x/tools v0.0.0-20201013201025-64a9e34f3752
[2020-10-19T14:33:42.637Z] go: downloading golang.org/x/mod v0.3.0
[2020-10-19T14:33:42.637Z] go: downloading golang.org/x/xerrors v0.0.0-20200804184101-5ec99f83aff1
[2020-10-19T14:33:42.637Z] go: golang.org/x/tools upgrade => v0.0.0-20201017001424-6003fad69a88
[2020-10-19T14:33:42.637Z] go: golang.org/x/mod upgrade => v0.3.0
[2020-10-19T14:33:42.637Z] go: golang.org/x/xerrors upgrade => v0.0.0-20200804184101-5ec99f83aff1
[2020-10-19T14:33:42.637Z] goimports -local github.com/elastic/elastic-package/ -w .
[2020-10-19T14:33:42.637Z] go get -u golang.org/x/lint/golint
[2020-10-19T14:33:42.637Z] go: downloading golang.org/x/lint v0.0.0-20200302205851-738671d3881b
[2020-10-19T14:33:42.637Z] go: found golang.org/x/lint/golint in golang.org/x/lint v0.0.0-20200302205851-738671d3881b
[2020-10-19T14:33:42.637Z] go: golang.org/x/tools upgrade => v0.0.0-20201017001424-6003fad69a88
[2020-10-19T14:33:44.025Z] go list ./... | xargs -n 1 golint -set_exit_status
[2020-10-19T14:33:45.413Z] /var/lib/jenkins/workspace/Beats_elastic-package_PR-143/src/github.com/elastic/elastic-package/internal/fields/validate.go:19:6: exported type Validator should have comment or be unexported
[2020-10-19T14:33:45.413Z] /var/lib/jenkins/workspace/Beats_elastic-package_PR-143/src/github.com/elastic/elastic-package/internal/fields/validate.go:23:1: exported function CreateValidatorForDataStream should have comment or be unexported
[2020-10-19T14:33:45.413Z] /var/lib/jenkins/workspace/Beats_elastic-package_PR-143/src/github.com/elastic/elastic-package/internal/fields/validate.go:50:1: exported method Validator.ValidateDocumentBody should have comment or be unexported
[2020-10-19T14:33:45.413Z] Found 3 lint suggestions; failing.
[2020-10-19T14:33:47.969Z] Makefile:15: recipe for target 'lint' failed
[2020-10-19T14:33:47.969Z] make: *** [lint] Error 123
[2020-10-19T14:33:48.279Z] Running on Jenkins in /var/lib/jenkins/workspace/Beats_elastic-package_PR-143
[2020-10-19T14:33:48.402Z] [INFO] getVaultSecret: Getting secrets
[2020-10-19T14:33:48.451Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2020-10-19T14:33:49.434Z] + chmod 755 generate-build-data.sh
[2020-10-19T14:33:49.434Z] + ./generate-build-data.sh https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/elastic-package/PR-143/ https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/elastic-package/PR-143/runs/1 FAILURE 493680
[2020-10-19T14:33:49.434Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/elastic-package/PR-143/runs/1/steps/?limit=10000 -o steps-info.json
[2020-10-19T14:33:49.684Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/elastic-package/PR-143/runs/1/tests/?status=FAILED -o tests-errors.json
[2020-10-19T14:33:49.935Z] Retry 1/3 exited 22, retrying in 1 seconds...
[2020-10-19T14:33:51.278Z] Retry 2/3 exited 22, retrying in 2 seconds...
[2020-10-19T14:33:53.140Z] Retry 3/3 exited 22, no more retries left.
[2020-10-19T14:33:53.140Z] INFO: curl https://beats-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/Beats/elastic-package/PR-143/runs/1/log/ -o pipeline-log.txt

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

mtojek commented:
> I assume you will choose a different test package than aws when it comes to validating JSON results in system tests?

Yes, I will try to use the only one present, `apache`, but it would be nice to spend some time and populate the repo (already created such issue in integrations).

mtojek commented:
Not sure what should I do with entries like:

```
[2020-10-20T16:31:02.897Z] Error: error running package system tests: could not check for expected data in data stream: one or more errors found in documents stored in logs-apache.access-ep data stream: [0] field "agent.hostname" is not defined
```

The field `agent.hostname` is not defined anywhere as it's a common one. Should I accept it even though there is no definition for it in `fields.yml`?

ycombinator commented:
> The field `agent.hostname` is not defined anywhere as it's a common one. Should I accept it even though there is no definition for it in `fields.yml`?

I assume this field is somehow getting defined in the index template's mapping. Do you how where that's happening?

mtojek commented:
> I assume this field is somehow getting defined in the index template's mapping. Do you how/where that's happening?

No idea. Maybe @ruflin know some more details? I suspect that there are more hidden definitions.

mtojek commented:
The system test runner reports tests results under the "Tests" tab:
https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Felastic-package/detail/PR-143/12/tests

Moving on...

mtojek commented:
I need to fix the `geo_point` field:

```
[0] field "source.geo.location.lon" is not defined
[1] field "source.geo.location.lon" is not defined
[2] field "source.geo.location.lon" is not defined
[3] field "source.geo.location.lon" is not defined
[4] field "source.geo.location.lon" is not defined
[5] field "source.geo.location.lon" is not defined
[6] field "source.geo.location.lat" is not defined
[7] field "source.geo.location.lat" is not defined
```

ycombinator commented:
@mtojek Could you update the PR description with a short description of what this PR does and why? And/or link it to an issue? This will help if anyone stumbles upon this PR in the future via git blame or something to figure out why we did what we did here 😄 . Thanks!

ycombinator commented on Makefile:
```diff
@@ -22,13 +22,16 @@ licenser:
 	go mod tidy
 
+test-go:
+	go test -v -count 1 ./...
```
Consider adding an explanation for the `-count 1`, similar to https://github.com/elastic/package-spec/blob/e7102f8cbd392d977bb2a21b12674dbe2ae8b985/code/go/Makefile#L33-L35.

mtojek replied:
Fixed

ycombinator commented on internal/fields/validate.go:
```diff
@@ -0,0 +1,196 @@
+// CreateValidatorForDataStream method creates a validator for the data stream.
+func CreateValidatorForDataStream(dataStreamRootPath string) (*Validator, error) {
+	fieldsDir := filepath.Join(dataStreamRootPath, "fields")
+	fis, err := ioutil.ReadDir(fieldsDir)
```
Nit: It's not clear to me what `fis` stand for. But I understand that it will contain a list of fields files. Maybe rename the variable to `fieldsFiles` instead of `fis` to be a bit more descriptive and to parallel `fieldsDir`?

mtojek replied:
This is actually `fileInfos`, but I agree it's not really descriptive. Renamed!

ycombinator commented on internal/fields/validate.go:
```diff
@@ -0,0 +1,196 @@
+	return nil
+}
+
+func (v *Validator) validateElementFormat(key string, val interface{}) error {
```
Nit: consider renaming this method to `validatScalarElement` or similar so it parallel's `validateMapElement`.

mtojek replied:
Renamed. Thanks!

ycombinator commented on internal/fields/validate.go:
```diff
@@ -0,0 +1,196 @@
+	return nil
+}
+
+func skipValidationForField(key string) bool {
```
At the very least we should add a comment here explaining why these fields are being skipped during validation. Additionally, should we file an issue to follow up on some way to validate these in the future (related to the discussion started here: https://github.com/elastic/elastic-package/pull/143#issuecomment-712994535)?

mtojek replied:
Created an issue: https://github.com/elastic/elastic-package/issues/147

ycombinator commented on internal/fields/validate.go:
```diff
@@ -0,0 +1,196 @@
+	return key == family || strings.HasPrefix(key, family+".")
+}
+
+func findElementDefinitionInSlice(root, searchedKey string, fieldDefinitions []fieldDefinition) *fieldDefinition {
```
Nit: consider renaming `Slice` to something else that is more meaningful to the problem domain like `FieldDefinitions`. `Slice` can be inferred by looking at the data type of the last argument anyway.

mtojek replied:
You're right here. Renamed to `findElementDefinition`.

ycombinator commented on internal/fields/validate.go:
```diff
@@ -0,0 +1,196 @@
+			continue
+		}
+
+		fd := findElementDefinitionInSlice(key, searchedKey, def.Fields)
```
To avoid star recursion here, I wonder if we should store the field definitions in a flattened manner. Then we can just iterated over the flattened list and check prefixes or do some sort of glob matching instead of recursion. Could be an optimization for a follow up PR.

mtojek replied:
I don't remember the exact thought process now, but I had a similar discussion with myself.

 Let's consider a field in results: `apache.access.foo.bar`. If we flatten field definitions (from YML files), we'll end up with 4 records to cover the single result fields -  we need field definitions for `apache`, `apache.access`, `apache.access.foo`, `apache.access.foo.bar`.

I considered this as not a huge gain comparing to the recursive solution (more towards readability?), but I'm happy to adjust it in follow-ups.

ycombinator replied:
++ to leaving it this way for now. Only reason to try the flattening ideal would be for performance / lower memory usage but until either of those things become an issue, let's not optimize prematurely.

ycombinator commented on internal/fields/validate.go:
```diff
@@ -0,0 +1,196 @@
+	k := strings.ReplaceAll(key, ".", "\\.")
+	k = strings.ReplaceAll(k, "*", "[^.]+")
+
+	// Workaround for potential geo_point, as "lot" and "lat" fields are not present in field definitions.
```
Typo: should be `lon`.

mtojek replied:
Fixed

ycombinator commented on internal/fields/validate.go:
```diff
@@ -0,0 +1,196 @@
+	k = fmt.Sprintf("^%s$", k)
+	matched, err := regexp.MatchString(k, searchedKey)
+	if err != nil {
+		panic(err)
```
Since we're `panic`ing here, maybe add a bit more context with `errors.Wrap`?

mtojek replied:
Fixed.

ycombinator commented on internal/fields/validate.go:
```diff
@@ -0,0 +1,196 @@
+	return nil
+}
+
+func ensureSingleElementValue(val interface{}) (interface{}, bool) {
```
By "single" do you mean "scalar"?

ycombinator commented on internal/fields/validate.go:
```diff
@@ -0,0 +1,196 @@
+	return nil
+}
+
+func ensureSingleElementValue(val interface{}) (interface{}, bool) {
```
Took me a few seconds to realize why we need this helper: because any data type in ES can take array values. Maybe add an explanatory comment here for future us?

mtojek replied:
I added a comment.

ycombinator commented on internal/multierror/error.go:
```diff
@@ -0,0 +1,26 @@
+		strs[i] = fmt.Sprintf("[%d] %v", i, err)
+	}
+	return strings.Join(strs, "\n")
+}
```
❤️

ycombinator commented on internal/testrunner/runners/system/runner.go:
```diff
@@ -247,26 +262,32 @@ func (r *runner) run() ([]testrunner.TestResult, error) {
+
+			err := fieldsValidator.ValidateDocumentMap(hit.Source)
+			if err != nil {
+				multiErr = append(multiErr, err)
```
Consider adding a `continue` here. I know it's not needed because this is the last check in the loop but  it future-proofs the code in case we add more code to the loop after this (and forget to add the `continue` here at the time).

mtojek replied:
This is great argument, really appreciate it. Introduced of course.

ycombinator commented on internal/testrunner/runners/system/runner.go:
```diff
@@ -247,26 +262,32 @@ func (r *runner) run() ([]testrunner.TestResult, error) {
-
 	if err != nil {
-		return resultsWith(result, errors.Wrap(err, "could not check for expected data in data stream"))
+		return resultsWith(result, err)
```
Any reason to review the `errors.Wrap` here?

mtojek replied:
It is like that because I changed the implementation of `resultsWith` function:

```go
	resultsWith := func(tr testrunner.TestResult, err error) ([]testrunner.TestResult, error) {
		tr.TimeElapsed = time.Now().Sub(startTime)
		if err == nil {
			return []testrunner.TestResult{tr}, nil
		}

		if tcf, ok := err.(testerrors.ErrTestCaseFailed); ok {
			tr.FailureMsg = tcf.Reason
			tr.FailureDetails = tcf.Details
			return []testrunner.TestResult{tr}, nil
		}

		tr.ErrorMsg = err.Error()
		return []testrunner.TestResult{tr}, err
	}
```

You can say, that I'm doing type assertion: `err.(testerrors.ErrTestCaseFailed)`.

ycombinator replied:
Ah, I see now. Thanks for pointing me to that.

ycombinator reviewed (commented):
Really nice work on this! Most of my comments are minor.

ycombinator commented:
> I added a couple of fields and it seems that there're new weird ones coming :) (log.input?). Maybe we shouldn't fail on undefined fields and just check the type correctness? I see now a case in which a field change pushed to filebeat/metricbeat (new field added) fails all system tests for integrations.

I imagine such new common fields should not be added frequently so I would actually prefer to be strict, at least initially and see how much of a PITA it becomes for us. If we suddenly see *all* tests failing for integrations, it's a pretty good indication that it's not something wrong with one or two integrations but something more widespread and common. Looking at the failure details it should become pretty clear what that common new field is and then we can add it to our skip list.

Also, depending on how we solve #147, we might not have to worry about this issue of new common fields being added every so often.

mtojek commented:
Ok, I will adjust the implementation to report all problems and then add missing fields. I'm only worried about future situations in which we have to adjust all integrations (time consuming), but maybe it won't be so hard.

ycombinator commented:
> Ok, I will adjust the implementation to report all problems and then add missing fields. I'm only worried about future situations in which we have to adjust all integrations (time consuming), but maybe it won't be so hard.

Yes, this is a fair point. I think we'll have to see how it goes. If it starts to become a hassle, I'll be +1 to not failing on undefined fields. But I'm also hoping we can find an elegant solution to #147, something that involves being able to detect these common fields from some other source than the package field definition files.

mtojek commented:
I modified the code to get all validation errors at once, but it went up to 300 errors :) See: https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Felastic-package/detail/PR-143/33/tests

I will adjust the code to report only unique records.

mtojek commented:
I managed to clean up all reported fields: https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Felastic-package/detail/PR-143/35/tests

Field families that are not related to the specific integration, but appear in documents: `cloud.*`, `host.*`, `event.*`.

mtojek commented:
@ycombinator I fixed all issues and addressed your comments. Please take a look when you're free. Keep in mind the number of additional fields I had to put here.

ycombinator commented on internal/fields/validate.go:
```diff
@@ -0,0 +1,217 @@
+		isFieldFamilyMatching("event", key) || // too many common fields
+		isFieldFamilyMatching("host", key) || // too many common fields
+		isFieldFamilyMatching("metricset", key) || // field is deprecated
+		isFieldFamilyMatching("event.module", key) // field is deprecated
```
This line is now redundant with `isFieldFamilyMatching("event", key)` but I think we should keep it anyway because even if we were to somehow get rid of `isFieldFamilyMatching("event", key)` in the future, we would want this line to be here, correct?

mtojek replied:
Right, maybe it's better to keep it here even though it's redundant.

ycombinator reviewed (approved):
LGTM. Nice job on this feature!
